### ISSUE TITLE
FLAKE E2E: [FAIL] MagicCatalog when an existing magic catalog exists [It] should succeed when the magic catalog is updated

### ISSUE THREAD

**Issue by tmshort** [operator-framework/operator-lifecycle-manager#3146]
## Flaky Test Report
This is occurring more and more frequently.

Summarizing 1 Failure:
  [FAIL] MagicCatalog when an existing magic catalog exists [It] should succeed when the magic catalog is updated
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/magic_catalog_test.go:71

**Failure Log Link**
[Failure Log](https://github.com/operator-framework/operator-lifecycle-manager/actions/runs/7410754771/job/20164628205?pr=3144#step:5:2232)
You can get the link by clicking on the line number in the job logs and copying the url from your browser.

**Relevant Failure Log**

```
------------------------------
• [FAILED] [322.714 seconds]
MagicCatalog when an existing magic catalog exists [It] should succeed when the magic catalog is updated
/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/magic_catalog_test.go:66

  Captured StdOut/StdErr Output >>
  Using the kubectl kubectl binary
  Using the artifacts/magic-catalog-e2e-78gnv output directory
  Storing the test artifact output in the artifacts/magic-catalog-e2e-78gnv directory
  Collecting get catalogsources -o yaml output...
  Collecting get subscriptions -o yaml output...
  Collecting get operatorgroups -o yaml output...
  Collecting get clusterserviceversions -o yaml output...
  Collecting get installplans -o yaml output...
  Collecting get pods -o wide output...
  No resources found in magic-catalog-e2e-78gnv namespace.
  Collecting get events --sort-by .lastTimestamp output...
  << Captured StdOut/StdErr Output

  Timeline >>
  created the magic-catalog-e2e-78gnv testing namespace
  created the magic-catalog-e2e-78gnv/magic-catalog-e2e-78gnv-operatorgroup operator group
  [FAILED] in [It] - /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/magic_catalog_test.go:71 @ 01/04/24 15:09:48.772
  collecting the magic-catalog-e2e-78gnv namespace artifacts as the 'should succeed when the magic catalog is updated' test case failed
  collecting logs in the ./artifacts/ artifacts directory
  tearing down the magic-catalog-e2e-78gnv namespace
  resetting e2e kube client
  deleting magic-catalog-e2e-78gnv/mc-e2e-catalog-p4p4r-pod
  deleting magic-catalog-e2e-78gnv/mc-e2e-catalog-p4p4r-configmap
  deleting magic-catalog-e2e-78gnv/mc-e2e-catalog-p4p4r
  deleting magic-catalog-e2e-78gnv/mc-e2e-catalog-p4p4r-svc
  deleting magic-catalog-e2e-78gnv/mc-e2e-catalog-p4p4r-pod
  deleting magic-catalog-e2e-78gnv/mc-e2e-catalog-p4p4r-configmap
  deleting magic-catalog-e2e-78gnv/magic-catalog-e2e-78gnv-operatorgroup
  deleting <global>/magic-catalog-e2e-78gnv
  garbage collecting CRDs
  << Timeline

  [FAILED] Expected
      <wait.errInterrupted>: 
      timed out waiting for the condition
      {
          cause: <*errors.errorString | 0xc0001b8110>{
              s: "timed out waiting for the condition",
          },
      }
  to be nil
  In [It] at: /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/magic_catalog_test.go:71 @ 01/04/24 15:09:48.772

  Full Stack Trace
    github.com/operator-framework/operator-lifecycle-manager/test/e2e.glob..func17.4.3()
    	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/magic_catalog_test.go:71 +0x323
------------------------------
```

**Comment by joelanford** [comment]
In the logs of the failed run, I found this:
```
2024-01-04T15:25:28.739907535Z stderr F time="2024-01-04T15:25:28Z" level=info msg="state.Key.Namespace=magic-catalog-e2e-d2mqw state.Key.Name=mc-e2e-catalog-b79bv state.State=CONNECTING"
2024-01-04T15:25:29.762994465Z stderr F time="2024-01-04T15:25:29Z" level=info msg="state.Key.Namespace=magic-catalog-e2e-d2mqw state.Key.Name=mc-e2e-catalog-b79bv state.State=TRANSIENT_FAILURE"
2024-01-04T15:25:39.419490022Z stderr F time="2024-01-04T15:25:39Z" level=info msg="state.Key.Namespace=magic-catalog-e2e-d2mqw state.Key.Name=mc-e2e-catalog-b79bv state.State=READY"
2024-01-04T15:25:39.547681038Z stderr F time="2024-01-04T15:25:39Z" level=info msg="state.Key.Namespace=magic-catalog-e2e-d2mqw state.Key.Name=mc-e2e-catalog-b79bv state.State=IDLE"
```

So it gets to READY, but then very quickly moves to IDLE.

I suspect this is the underlying issue. the state never moves back to READY. My hunch here is that there is a race condition between (a) this READY to IDLE transition and (b) the e2e checking for READY.
- if (a) wins, the test flakes
- if (b) wins, the test continues.

A quick fix here for this particular test might be to allow either READY or IDLE. 

A better fix may be to have the catalog operator perform its own health checks, which should keep its connection in the READY state.

**Comment by tmshort** [comment]
See discussion: https://kubernetes.slack.com/archives/C0181L6JYQ2/p1704381324922659

**Comment by joelanford** [comment]
Though it isn't clear to me why the connection would move so quickly from READY to IDLE. It seems like we need to better understanding of why that is happening before we decide on a more permanent fix.

**Comment by joelanford** [comment]
I did a few things:
1. I started up `opm serve` with the operatorhub catalog and using the same version of `opm` that it seems the e2e would use (latest). Then I ran a very short program that I wrote using the same grpc-go client library version that OLM is using that essentially just does `grpc.Dial` and then reports connection state changes. I ran it a few times, and each time the connection stayed in the READY state for the full 30m (the default idle timeout) before moving to IDLE.
2. I went dumpter diving to the GRPC client code that manages the connection state, and I don't see anything that stands out that would explain why the connection goes from READY to IDLE so quickly.

Next up, I'm going to try finding other examples of this flake to check to see if they all share this READY to IDLE transition. If so, I'll submit a PR that turns on GRPC logging/tracing so that we can hopefully get a better idea about what's going on.

**Comment by joelanford** [comment]
Ah, one more interesting observation: with my experiment from (1), I killed the `opm serve` command while the client was running and in the READY state. That caused the server to send a GOAWAY and close the connection. And _that_ caused the client to immediately move to the IDLE state.

After restarting the server, I noticed that my toy client did not attempt to re-establish the connection. I assume it won't until a subsequent RPC call is made.

Here's a hypothesis:
1. e2e test creates catalog.
1. Server starts, connection goes to READY, catalog operator successfully queries what it needs.
2. Server is killed (don't know the reason yet)
3. Catalog operator's GRPC client observes (2) and sets state to IDLE and updates catalog source status to IDLE
4. e2e test begins waiting for READY state to be observed in the catalog source status
5. e2e test timeout (5m) occurs prior to the catalog operator deciding to make another RPC (which would move the status from IDLE to READY)
